bin/cml.js

```diff
@@ -0,0 +1,68 @@
+#!/usr/bin/env node
+
+const { basename } = require('path');
+const { pseudoexec } = require('pseudoexec');
+
+const which = require('which');
+const winston = require('winston');
+const yargs = require('yargs');
+
+const configureLogger = (level) => {
+  winston.configure({
+    format: process.stdout.isTTY
+      ? winston.format.combine(
+          winston.format.colorize({ all: true }),
+          winston.format.simple()
+        )
+      : winston.format.json(),
+    transports: [
+      new winston.transports.Console({
+        handleExceptions: true,
+        handleRejections: true,
+        level
+      })
+    ]
+  });
+};
+
+const runPlugin = async ({ $0: executable, command }) => {
+  try {
+    if (command === undefined) throw new Error('no command');
+    const path = which.sync(`${basename(executable)}-${command}`);
+    const parameters = process.argv.slice(process.argv.indexOf(command) + 1); // HACK
+    process.exit(await pseudoexec(path, parameters));
```

Comment:
What the hell is this?? 😁 

> Node JS pseudoexec
Slightly inspired by node-kexec

🤯 😁

Reply:
😈  I can't escape your sharp eye! 🦅 

See also `exec` from [`man bash`](http://manpages.ubuntu.com/manpages/trusty/en/man1/bash.1.html) and [`exec(3)`](https://man7.org/linux/man-pages/man3/exec.3.html) for more information. It just forwards [`stdio(3)`](https://man7.org/linux/man-pages/man3/stdio.3.html) and [`syscalls(2)`](https://man7.org/linux/man-pages/man2/syscalls.2.html) to the child process, waits for it, and returns the exit code.

Node JS doesn't have anything similar to `exec` where you can replace your own process with another one, and `node-kexec` is just a stale project without multiplatform support. So... I had to create yet another ~odd job~ package implementing a similar functionality without native extensions.

Reply:
Can you explain to me whats the desired effect? It's it worth to create a new package?

Reply:
> _Can you explain to me whats the desired effect?_

Ideally, we should replace the `node` process interpreting `cml.js` with the plugin process, like what [`exec(3)`](https://man7.org/linux/man-pages/man3/exec.3.html) does.

That's not possible on Node JS, and the nearest alternative I have found is to [`spawn`](https://nodejs.org/api/child_process.html#child_process_child_process_spawn_command_args_options) the plugin process, [`inherit`](https://nodejs.org/api/child_process.html#child_process_options_stdio) the standard input/output/error and forward all the [signal events](https://nodejs.org/api/process.html#process_signal_events) to the child process.

> _It's it worth to create a new package?_

Yes, because it:
1. Is already published on `npm` 😉 
2. Resorts to some internal trickery that I would rather keep out of this project
3. May grow in the future if Node JS chooses not to expose the signal list anymore

Reply:
> Resorts to some internal trickery that I would rather keep out of this project

Sounds secure 😁

Reply:
@0x2b3bfa0 why have you removed the test files? They can help us to test that the commands work as the did.

Reply:
@DavidGOrtega, as far as I remember, I haven't removed any test file, but Git detects them as deleted/created because they have a few changes here and there hand have been renamed.

Reply:
I think that you have created new tests this is not the same:

```js
await exec(`node ./bin/cml.js publish --help`);
```

than this

```js
await exec(`node ./bin/cml-publish.js --help`);
```

Reply:
> _Resorts to some internal trickery that I would rather keep out of this project._

More explicitly, we use [`process.binding()`]() (which is [pending deprecation](https://nodejs.org/api/deprecations.html#DEP0111)) to [get a list of valid signals](https://github.com/0x2b3bfa0/node-pseudoexec/blob/7411ecccb83e1da1cee70e2bdd14a79f27604ee5/index.js#L5) for the current operating system, [exactly as they do](https://github.com/nodejs/node/blob/73d5f8a843c89707313eb2cf2b99af59096b738a/lib/internal/util.js#L44). 

> _Sounds secure_ 😁

`<kidding>` Do I look like somebody able to pull a supply chain attack on CML? Clue: I am. 😉  `</kidding>`

Reply:
> I think that you have created new tests this is not the same:
> 
> ```js
> await exec(`node ./bin/cml.js publish --help`);
> ```
> 
> than this
> 
> ```js
> await exec(`node ./bin/cml-publish.js --help`);
> ```
> With those test we would warranty that the old commands still works

Old commands will still work as soon as nobody touches `./bin/legacy.js` but we can't preserve the full path invocations.

Reply:
true, can we add a small test in the CI? Testing the old cli?

Reply:
Can we perhaps run `npx --package=. cml-publish --help` instead of `node ./bin/cml-publish.js --help` (?) Not sure if it's the best approach, though.

Reply:
@DavidGOrtega, please check  https://github.com/iterative/cml/pull/703/commits/b72aa269c35582776fbaa414c1e55fe50f45fc77

Reply:
Very nice!!  while we did not only have templating tests maybe, it would suffice at least checking that the command is accesible and at least generating the same output in one case

Reply:
> _What the hell is this??_ 😁

See also the [`pseudoexec`](https://www.npmjs.com/package/pseudoexec) package.